// @redwoodjs/cli/dist/commands/experimental/studio.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.EXPERIMENTAL_TOPIC_ID = void 0;
exports.builder = builder;
exports.description = exports.command = void 0;
exports.handler = handler;
var _cliHelpers = require("@redwoodjs/cli-helpers");
var _util = require("./util");
const command = exports.command = 'studio';
const description = exports.description = 'Run the Redwood development studio';
const EXPERIMENTAL_TOPIC_ID = exports.EXPERIMENTAL_TOPIC_ID = 4771;
function builder(yargs) {
  yargs.option('open', {
    default: true,
    description: 'Open the studio in your browser'
  }).epilogue((0, _util.getEpilogue)(command, description, EXPERIMENTAL_TOPIC_ID, true));
}
async function handler(options) {
  (0, _cliHelpers.recordTelemetryAttributes)({
    command: 'experimental studio',
    open: options.open
  });
  const {
    handler
  } = await import('./studioHandler.js');
  return handler(options);
}
